crypto/ec/ec_pmeth.c

```diff
@@ -121,8 +125,22 @@ static int pkey_ec_sign(EVP_PKEY_CTX *ctx, unsigned char *sig, size_t *siglen,
     }
 
     type = (dctx->md != NULL) ? EVP_MD_type(dctx->md) : NID_sha1;
-
-    ret = ECDSA_sign(type, tbs, tbslen, sig, &sltmp, ec);
+	
+    switch (dctx->nonce_type) {
+    case EVP_PKEY_ECDSA_NONCE_RANDOM:
+        ret = ECDSA_sign(0, tbs, tbslen, sig, &sltmp, ec);
+        break;
+    case EVP_PKEY_ECDSA_NONCE_DETERMINISTIC:
+        if (dctx->md == NULL) {
+            ECerr(EC_F_PKEY_EC_SIGN, EC_R_INVALID_DIGEST);
+            return 0;
+        }
+        ret = ec->meth->sign(type, tbs, tbslen, sig, &sltmp, NULL, NULL, ec);
```

Comment:
Should this be accomplished by calling
```
int ECDSA_sign_ex(int type, const unsigned char *dgst, int dgstlen,
                   unsigned char *sig, unsigned int *siglen,
                   const BIGNUM *kinv, const BIGNUM *rp, EC_KEY *eckey);
```

instead?

Reply:
An idea is to support deterministic ECDSA only at the EVP level, so that no code change would be required in the legacy ECDSA interface: you should only need to generate `kinv` and `rp` deterministically here, and then pass them down to `ECDSA_sign_ex()`.

Reply:
@romen The legacy behavior will not change.
eckey->meth->sign is not a public method, it will only be called from public method ECDSA_sign_ex.
And I've modified ECDSA_sign_ex, to pass zero as `type` to eckey->meth->sign, so that legacy behavior won't change.

Reply:
So if deterministic ECDSA is not achievable from the `ECDSA` layer public functions, wouldn't it be better to avoid unnecessary changes there and keep them all neatly contained in the EVP layer?

Reply:
@romen Okay, I will change it.

Reply:
But what about if this eventually will be FIPS approved thing and we will want to have it in the FIPS provider as well, wouldn't it complicate things?

Reply:
>> An idea is to support deterministic ECDSA only at the EVP level, so that no code change would be required in the legacy ECDSA interface: you should only need to generate `kinv` and `rp` deterministically here, and then pass them down to `ECDSA_sign_ex()`.
>>
>> if deterministic ECDSA is not achievable from the `ECDSA` layer public functions, wouldn't it be better to avoid unnecessary changes there and keep them all neatly contained in the EVP layer?
>>
> But what about if this eventually will be FIPS approved thing and we will want to have it in the FIPS provider as well, wouldn't it complicate things?

@t8m I am not sure about that, maybe @mattcaswell or @levitte can provide an answer on this?


If having the deterministic behaviour all contained at the `EVP` layer is not a problem for providers, I would try to push for reverting all the changes in `crypto/ec/ecdsa_ossl.c` in favor of changing L138 

https://github.com/openssl/openssl/blob/fd3c1a3534a439ea581114e2bf73bad59de40dd6/crypto/ec/ec_pmeth.c#L133-L139

with
1. a call to a function that generates the deterministic nonce and "returns" `kinv` and `r` from it, followed by
2. a call to 
   ```
   int ECDSA_sign_ex(int type, const unsigned char *dgst, int dgstlen,
                      unsigned char *sig, unsigned int *siglen,
                      const BIGNUM *kinv, const BIGNUM *rp, EC_KEY *eckey);
   ```
   in order to avoid any duplication of ECDSA-related code.

Reply:
I would note that currently with 1.1.1, if an ENGINE wanted to provide deterministic nonce generation for ECDSA without duplicating the ECDSA code unrelated to nonce generation, its only option would be through this kind of scheme where you generate the nonce, `kinv` and `r` and then call `ECDSA_sign_ex()`.

I feel it's already quite awkward that the internal
https://github.com/openssl/openssl/blob/97ace46e11dba4c4c2b7cb67140b6ec152cfaaf4/crypto/ec/ecdsa_ossl.c#L33-L35
diverged from the public facing `sign_setup` from
https://github.com/openssl/openssl/blob/97ace46e11dba4c4c2b7cb67140b6ec152cfaaf4/include/openssl/ec.h#L1298-L1310
and that
https://github.com/openssl/openssl/blob/97ace46e11dba4c4c2b7cb67140b6ec152cfaaf4/crypto/ec/ecdsa_ossl.c#L150-L152
calls directly the internal static `ecdsa_sign_setup` rather than using the `kmeth->sign_setup` one.

It's quite messy to describe the whole process and counterintuitive for the users of the ENGINE API for `EC_KEY_METHOD`, to the point that setting a custom `sign_setup` alone in case that's the only part that the ENGINE wants to change doesn't really work.

This is the reason why I am trying to resist going further down this road of making the internal `sign_setup` signature diverge from the public `sign_setup` signature.

Reply:
@romen You mean that we should add another new kmeth->sign_setup that matches the internal prototype and use the new one to initialize the kinv and r?
I don't understand why this is related to ENGINE?

Reply:
@JemmyLoveJenny apologies for being confusing!

The whole https://github.com/openssl/openssl/pull/9223#discussion_r300506501 was just trying to be an explanation of why I would try to avoid changes to `ecdsa_ossl.c` at all.

> I don't understand why this is related to ENGINE?

It's related because in the public API we don't only have
- https://github.com/openssl/openssl/blob/97ace46e11dba4c4c2b7cb67140b6ec152cfaaf4/include/openssl/ec.h#L1221
- https://github.com/openssl/openssl/blob/97ace46e11dba4c4c2b7cb67140b6ec152cfaaf4/include/openssl/ec.h#L1233
- https://github.com/openssl/openssl/blob/97ace46e11dba4c4c2b7cb67140b6ec152cfaaf4/include/openssl/ec.h#L1249-L1251  

but also 
- https://github.com/openssl/openssl/blob/97ace46e11dba4c4c2b7cb67140b6ec152cfaaf4/include/openssl/ec.h#L1298-L1310 mainly meant for `ENGINE`s (and soon providers) to provide alternative ways of doing signatures with `EC_KEY` objects.

Diverging 
https://github.com/openssl/openssl/blob/97ace46e11dba4c4c2b7cb67140b6ec152cfaaf4/crypto/ec/ecdsa_ossl.c#L33-L35 from https://github.com/openssl/openssl/blob/97ace46e11dba4c4c2b7cb67140b6ec152cfaaf4/include/openssl/ec.h#L1221 increases the functionality that is not directly reachable from the ECDSA public API and that an ENGINE/PROVIDER has an hard time to reimplement.

Again, all of this was just motivation of why I would avoid changes to `ecdsa_ossl.c` at all in this PR, if possible.

----

> You mean that we should add another new `kmeth->sign_setup` that matches the internal prototype and use the new one to initialize the `kinv` and `r`?

No, this is the change I would like to have before approving this PR:

> If having the deterministic behaviour all contained at the `EVP` layer is not a problem for providers, I would try to push for reverting all the changes in `crypto/ec/ecdsa_ossl.c` in favor of changing L138 _from_
> 
> https://github.com/openssl/openssl/blob/fd3c1a3534a439ea581114e2bf73bad59de40dd6/crypto/ec/ec_pmeth.c#L133-L139
> 
> with
> 
> 1. a call to a function that generates the deterministic nonce and "returns" `kinv` and `r` from it, followed by
> 2. a call to
>    ```
>    int ECDSA_sign_ex(int type, const unsigned char *dgst, int dgstlen,
>                       unsigned char *sig, unsigned int *siglen,
>                       const BIGNUM *kinv, const BIGNUM *rp, EC_KEY *eckey);
>    ```
>    
>    
>    in order to avoid any duplication of ECDSA-related code.